### PR TITLE
Configuration refactoring

### LINKED ISSUE #413: Network type selectable at compile time
Network type (e.g. mainnet/testnet) should be selectable at compile time with an option and select the appropriate set of variables.

### PR DESCRIPTION
**Pull request by thibault-martinez**
- Network type is now a compile time option `--define network=mainnet` or `--define network=testnet` (defaults to `mainnet`)
- Set some values at compile time
- Removes all occurrences of the `testnet` variables
- Split configurations in 4 conf objects `iota_ciri_conf_t`, `iota_gossip_conf_t`, `iota_consensus_conf_t` and `iota_api_conf_t` (not in the PR yet)
- Configuration cycle : default -> file -> CLI
- All the configurations variables will be settable by CLI
- Fixes #413 

# Test Plan:
CI must pass.

### CODE REVIEW

File: ciri/core.c [hunk 1]
```diff
@@ -25,29 +25,22 @@ retcode_t core_init(core_t* const core) {
...
-  if (requester_init(&core->transaction_requester, &core->consensus.tangle)) {
-    log_critical(CORE_LOGGER_ID, "Initializing transaction requester failed\n");
-    return RC_CORE_FAILED_REQUESTER_INIT;
-  }
-
-  log_info(CORE_LOGGER_ID, "Initializing cIRI node\n");
-  if (node_init(&core->node, core, &core->consensus.tangle,
-                &core->transaction_requester)) {
-    log_critical(CORE_LOGGER_ID, "Initializing cIRI node failed\n");
+  log_info(CORE_LOGGER_ID, "Initializing gossiping node\n");
```
Comment (tsvisabo) [inline ciri/core.c]: can you change `Initializing gossiping node` -> `Initializing node gossip components` , it's sounds as if the node can't keep a secret.

**Review by tsvisabo — CHANGES_REQUESTED** [review]

File: ciri/core.c (hunk 1 above)
Comment (thibault-martinez) [reply ciri/core.c]: Done

**Review by tsvisabo — APPROVED** [review]

File: ciri/api/conf.c [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+#include "ciri/api/conf.h"
+
+retcode_t iota_api_conf_init(iota_api_conf_t* const conf) {
+  if (conf == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  conf->port = DEFAULT_API_PORT;
+
+  // Limits conf
```
Comment (th0br0) [inline ciri/api/conf.c]: Add an issue?

File: ciri/api/tests/test_broadcast_transactions.c [hunk 3]
```diff
@@ -98,10 +98,11 @@ int main(void) {
   api.transaction_validator = &transaction_validator;
   broadcaster_init(&broadcaster, &node);
   api.broadcaster = &broadcaster;
-  memset(defs.genesis_hash, FLEX_TRIT_NULL_VALUE, FLEX_TRIT_SIZE_243);
+  TEST_ASSERT(iota_consensus_conf_init(&conf) == RC_OK);
+  conf.snapshot_timestamp_sec = 1536845195;
+  conf.mwm = 10;
```
Comment (th0br0) [inline ciri/api/tests/test_broadcast_transactions.c]: for tests we should have a low mwm?

File: ciri/conf.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct iota_ciri_conf_s {
+  logger_level_t log_level;
+} iota_ciri_conf_t;
+
+retcode_t iota_ciri_conf_default(iota_ciri_conf_t *const ciri_conf,
```
Comment (th0br0) [inline ciri/conf.h:44]: Document the method please :upside_down_face:

File: ciri/conf.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+                                 iota_consensus_conf_t *const consensus_conf,
+                                 iota_gossip_conf_t *const gossip_conf,
+                                 iota_api_conf_t *const api_conf);
+retcode_t iota_ciri_conf_cli(iota_ciri_conf_t *const ciri_conf,
```
Comment (th0br0) [inline ciri/conf.h:77]: Document the method please :upside_down_face:

File: consensus/conf.h [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+// Milestone conf
+
+#define DEFAULT_NUM_KEYS_IN_MILESTONE NUM_KEYS_IN_MILESTONE
+#define DEFAULT_MWN MWM
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct iota_consensus_conf_s {
```
Comment (th0br0) [inline consensus/conf.h:33]: Documentation? :upside_down_face:

File: consensus/conf.h [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  uint64_t snapshot_signature_depth;
+  flex_trit_t snapshot_signature_pubkey[FLEX_TRIT_SIZE_243];
+  // Milestone conf
+  flex_trit_t coordinator[FLEX_TRIT_SIZE_243];
+  uint64_t last_milestone;
+  size_t num_keys_in_milestone;
+  uint8_t mwm;
+} iota_consensus_conf_t;
+
+retcode_t iota_consensus_conf_init(iota_consensus_conf_t* const conf);
```
Comment (th0br0) [inline consensus/conf.h:82]: Document the method please :upside_down_face:

File: gossip/conf.h [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+// Probabilities conf
+
+#define DEFAULT_PROBABILITY_SELECT_MILESTONE_CHILD 0.7
+#define DEFAULT_PROBABILITY_REMOVE_REQUEST 0.01
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct iota_gossip_conf_s {
```
Comment (th0br0) [inline gossip/conf.h:31]: Document at least the probabilities?

**Review by th0br0 — CHANGES_REQUESTED** [review]
just documentation nits... :)

File: ciri/api/tests/test_broadcast_transactions.c (hunk 3 above)
Comment (thibault-martinez) [reply ciri/api/tests/test_broadcast_transactions.c]: Sure

**Review by th0br0 — APPROVED** [review]

**Pull request merged**